recipes-extended/xen/files/libxl-xenmgr-support.patch

```diff
@@ -108,31 +111,75 @@ PATCHES
          ret = ERROR_INVAL;
 --- a/tools/libxl/libxl_dm.c
 +++ b/tools/libxl/libxl_dm.c
-@@ -2353,6 +2353,24 @@ retry_transaction:
+@@ -2109,6 +2109,9 @@ static int libxl__store_libxl_entry(libx
+     return libxl__xs_printf(gc, XBT_NULL, path, "%s", value);
+ }
+ 
++static void argo_xswait_cb(libxl__egc *egc, libxl__xswait_state *xswait,
++                              int rc, const char *p);
++
+ static void spawn_stubdom_pvqemu_cb(libxl__egc *egc,
+                                 libxl__dm_spawn_state *stubdom_dmss,
+                                 int rc);
+@@ -2353,11 +2356,22 @@ retry_transaction:
      libxl__xs_printf(gc, XBT_NULL,
                       DEVICE_MODEL_XS_PATH(gc, dm_domid, guest_domid, "/xen_extended_power_mgmt"), "2");
  
+-    libxl__multidev_begin(ao, &sdss->multidev);
+-    sdss->multidev.callback = spawn_stub_launch_dm;
+-    /* OpenXT: Again, no disk for the stubdom itself */
+-    libxl__add_disks(egc, ao, dm_domid, dm_config, &sdss->multidev);
+-    libxl__multidev_prepared(egc, &sdss->multidev, 0);
 +    /* OpenXT: We wait on xenmgr writing argo-firewall-ready.
 +     *         xenmgr waits on us writing image/device-model-domid
 +     */
-+    unsigned int timeout;
-+    char *ready = NULL;
 +    const char *firewall_path =
 +        GCSPRINTF("%s/argo-firewall-ready",
 +                  libxl__xs_get_dompath(gc, guest_domid));
-+    /* Block and wait for argo firewall rules */
-+    for (timeout = 0; !ready && timeout < 4; ++timeout) {
-+        ready = xs_read(ctx->xsh, XBT_NULL, firewall_path, NULL);
-+        if (ready) {
-+            free(ready);
-+            break;
-+        }
-+        sleep(1);
++    /* Wait for argo firewall rules */
++    libxl__xswait_init(&sdss->xswait);
++    sdss->xswait.ao = ao;
++    sdss->xswait.what = GCSPRINTF("Domain %u for argo-firewall-ready",
++                                  guest_domid);
++    sdss->xswait.path = firewall_path;
++    sdss->xswait.timeout_ms = LIBXL_ARGO_TIMEOUT * 1000;
++    sdss->xswait.callback = argo_xswait_cb;
++    ret = libxl__xswait_start(gc, &sdss->xswait);
++    if (ret) goto out;
+ 
+     return;
+ 
+@@ -2366,6 +2380,30 @@ out:
+     spawn_stubdom_pvqemu_cb(egc, &sdss->pvqemu, ret);
+ }
+ 
++static void argo_xswait_cb(libxl__egc *egc, libxl__xswait_state *xswait,
++                           int rc, const char *p)
```

Comment:
p here is the value of the watched xenstore node.  Do you want to verify it's the expected value?

Reply:
Yes, I don't see any reason why not to verify p's value. However, I don't think there's anything better to do than to log an error message and continue on our merry way, same as in the case of a timeout or other error at this point.